pkg/kine/prepared/db.go

```diff
@@ -0,0 +1,191 @@
+package prepared
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"sync"
+	"sync/atomic"
+)
+
+type DB struct {
+	underlying *sql.DB
+	mu         sync.Mutex
+	maxSize    int
+	store      map[string]*lruEntry
+	lruList    lruList
+}
+
+func New(db *sql.DB, maxSize int) *DB {
+	return &DB{
+		underlying: db,
+		maxSize:    maxSize,
+		store:      make(map[string]*lruEntry, maxSize),
+	}
+}
+
+func (db *DB) Underlying() *sql.DB { return db.underlying }
+
+func (db *DB) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	stms, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stms.ExecContext(ctx, args...)
+}
+
+func (db *DB) QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
+	stms, err := db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	return stms.QueryContext(ctx, args...)
+}
+
+func (db *DB) Close() error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	errs := []error{}
+	for len(db.store) != 0 {
+		if err := db.drop(db.lruList.tail()); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := db.underlying.Close(); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (db *DB) touch(entry *lruEntry) {
+	db.lruList.remove(entry)
+	db.lruList.add(entry)
+}
+
+func (db *DB) prepare(ctx context.Context, query string) (*stmt, error) {
+	if stmt := db.get(query); stmt != nil {
+		return stmt, nil
+	}
+	prepared, err := db.underlying.PrepareContext(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	result := &stmt{prepared, atomic.Int32{}}
+	result.Ref()
```

Comment:
The need to immediately add one to the reference counter is quite fragile, I think in this case, it would be better to have a function which does the value construction for us, perhaps `newStmt(prepared)`?

It also looks very suspicious that the result of `.Ref()` is being ignored, even if it's okay in this case. Also, I think this call to increment the reference counter (if kept) should go immediately above the `return result` line as `db.put` already increments it internally so this would make it clearer what the refCount++ is referring to